packages/providers/src/transaction-response/transaction-response.ts

```diff
@@ -200,10 +200,22 @@ export class TransactionResponse {
   async waitForResult<TTransactionType = void>(
     contractsAbiMap?: AbiMap
   ): Promise<TransactionResult<TTransactionType>> {
-    for await (const { statusChange } of this.provider.operations.statusChange({
+    let transactionCompleted = false;
+
+    // Check if the transaction is already completed
+    const { transaction } = await this.provider.operations.getTransaction({
       transactionId: this.id,
-    })) {
-      if (statusChange.__typename !== 'SubmittedStatus') break;
+    });
+    if (transaction?.status?.type !== 'SubmittedStatus') {
+      transactionCompleted = true;
+    }
+
+    if (!transactionCompleted) {
+      for await (const { statusChange } of this.provider.operations.statusChange({
+        transactionId: this.id,
+      })) {
+        if (statusChange.__typename !== 'SubmittedStatus') break;
+      }
```

Comment:
I've set a local node that takes 12 seconds to process transactions and placed a log inside this `for` loop. 

It seems `statusChange` does not have the `__typename` property, therefore is always `undefined`.

The `statusChange` is being logged this way:

```ts
{ type: 'SubmittedStatus', time: '4611686020123395412' }
```

and after changing statuses

```ts
{
  type: 'SuccessStatus',
  block: {
    id: '0x03565ff5aec3c976d23ba6d87f63cf586de47b14b3aff8115e85ad19f0008579'
  },
  time: '4611686020123395418',
  programState: { returnType: 'RETURN', data: '0x0000000000000000' }
}
```
I believe this loop always `breaks` when reaching this `if` statement.

Reply:
Great find, @Torres-ssf! I thought due to all tests passing that the functionality was correct, but it seems like the bug in the loop didn't make any difference because the transactions in the tests were always successful (and speedy - key problem). I pushed a bugfix in #1304 for this as well. #1281 would've probably shown this as well. Good that you caught it earlier.

Although this is a great find, I don't think this is the root cause of this issue, as I've explained [here](https://github.com/FuelLabs/fuels-ts/issues/1301#issuecomment-1741185766).